### ISSUE TITLE
Line Endings Are Always CRLF on Windows

### ISSUE THREAD

**Issue by Kurt-von-Laven** [frnmst/md-toc#33]
According to the documentation, `--newline-string` is:
> the string used to separate the lines of the TOC. Use quotes to delimit the string. Defaults to `'\n'`

However, on Windows, `--newline-string` actually defaults to `'\r\n'`, which is desirable and simply needs to be documented. The larger problem is that setting `--newline-string '\n'` has no effect.

Consequently, if you run [the mixed-line-ending hook](https://github.com/pre-commit/pre-commit-hooks#mixed-line-ending) with `--fix=lf` after md-toc, it will fail if md-toc was fed any Markdown files containing `<!--TOC-->`. This remains true no matter how many times you run the hooks, which can be disorienting since mixed-line-ending outputs "fixed mixed line endings." To facilitate debugging, pre-commit hooks should fail whenever they modify any file, even if only to change the line endings.

I am on version 8.0.1 of md-toc.

**Comment by frnmst** [comment]
Hello,

Somehow windows  (or Python on Windows) ignores my decision of using `\n` as default line terminator. See:

- https://github.com/frnmst/md-toc/blob/master/md_toc/constants.py#L29
- https://github.com/frnmst/md-toc/blob/master/md_toc/cli.py#L296

If I run md-toc with `--newline-string` on linux and then i run `file` on the generated file I get:

- `\r\n`: `ASCII text, with CRLF line terminators`
- `\r`: `ASCII text, with CR line terminators`
- `\n`: `ASCII text`

As a side note, I never tested md-toc on windows and It's been 10 years since I last used windows so I can't do any debug.

**Comment by Kurt-von-Laven** [comment]
The bug is in [fpyutils.filelines.insert_string_at_line](https://docs.franco.net.eu.org/fpyutils/api.html#fpyutils.filelines.insert_string_at_line). When the file is [opened for writing](https://software.franco.net.eu.org/frnmst/fpyutils/src/branch/master/fpyutils/filelines.py#L67), `newline` is not specified and so [defaults to None](https://docs.python.org/3/library/functions.html#open). Hence, the file is [opened in universal newline mode](https://docs.python.org/3/library/functions.html#open-newline-parameter):
> When writing output to the stream, if `newline` is `None`, any `'\n'` characters written are translated to the system default line separator, [os.linesep](https://docs.python.org/3/library/os.html#os.linesep).

I noted that `newline_character` [wasn't passed](https://github.com/frnmst/md-toc/blob/master/md_toc/api.py#L72-L77) to `insert_string_at_line()`. One solution would be to both pass `newline_string` as `newline_character` there and also pass `newline_character` to `open()` in `insert_string_at_line()` in `write_string_on_file_between_markers()`. What would you suggest? Happy to open a pull request.

**Comment by frnmst** [comment]
Very interesting. In my case these are the results using `cat -e` on this example:

```markdown
# b

## a.mf

### j.gh

## as

# a
```
`pipenv run python -m md_toc --newline-string '\n' -p github -l6 a.md`

```
$
# b$
$
## a.mf$
$
### j.gh$
$
## as$
$
# a$
$
$
<!--TOC-->$
$
- [b](#b)$
  - [a.mf](#amf)$
    - [j.gh](#jgh)$
  - [as](#as)$
- [a](#a)$
$
<!--TOC-->$
```

`pipenv run python -m md_toc --newline-string '\r\n' -p github -l6 a.md`

```markdown
$
# b$
$
## a.mf$
$
### j.gh$
$
## as$
$
# a$
$
$
<!--TOC-->$
$
- [b](#b)^M$
  - [a.mf](#amf)^M$
    - [j.gh](#jgh)^M$
  - [as](#as)^M$
- [a](#a)$
$
<!--TOC-->$
```

`pipenv run python -m md_toc --newline-string '\r' -p github -l6 a.md`

```markdown
$
# b$
$
## a.mf$
$
### j.gh$
$
## as$
$
# a$
$
$
<!--TOC-->$
$
- [b](#b)^M  - [a.mf](#amf)^M    - [j.gh](#jgh)^M  - [as](#as)^M- [a](#a)$
$
<!--TOC-->$
```

By all means open a PR on the dev branch and on fpyutils dev branch.

**Comment by frnmst** [comment]
Should be fixed now in 8.1.0. Feel free to re-open if not working.
